docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+of X+1000. HVPA does not act, and the pod is permanently stuck starving for 1500 millicores of CPU.
+
+Lastly, HVPA applies additive model to the interaction between HPA and VPA, with respect to provided compute power, but
+that interaction is actually multiplicative. As an extreme example, take a workload running at one replica with one core.
+Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
+suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
+each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for
+the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 
+
+### Other Benefits to Replacing the Existing Solution
+- HVPA applies recommendations via direct edits to the pod objects. This approach requires nonobvious special handling
+  by Gardener code outside HVPA, which is error-prone.
+
+### Goals
+- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution
+- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
+  HVPA's policy for early horizontal scale-out
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi
+- Reduce Gardener custom code base and ongoing maintenance effort
+
+### Non-Goals
+- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope
+  of this GEP
+- Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of
+  this GEP
+
+## Proposal
+ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
+use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low)
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate
```

Comment:
As far as I'm aware, there is no monitoring stack deployed for clusters of purpose `testing` ([ref](https://github.com/gardener/gardener/blob/afe07938c905c47cef81565ebafc1c203dc96978/pkg/operation/operation.go#L576-L579)). But without a shoot prometheus, HPA scaling would not work. 

@istvanballok - would be great, if you could confirm/correct :) Thanks!

If that behavior is intended, it would be great to document it somewhere.

Reply:
> As far as I'm aware, there is no monitoring stack deployed for clusters of purpose testing ([ref](https://github.com/gardener/gardener/blob/afe07938c905c47cef81565ebafc1c203dc96978/pkg/operation/operation.go#L576-L579)). But without a shoot prometheus, HPA scaling would not work.

Thank you for catching this aspect @hendrikKahl: it is correct that the monitoring stack is currently not deployed into the control plane of shoot clusters that are labeled with the testing purpose.